File: src/data/typescript/barchart.ts
import { chromium, ChromiumBrowser, ElementHandle, Page } from "playwright";
import { upload_to_db_barchart } from "./database";

const industries = [
	"VURA",
	"VPTC",
	// ... rest of the industries
];

// This function is extract rows of data from a custom element on a webpage.
// In this case, the columns are "symbol" and "name" from barchart.com
async function extractSymbolAndName(
	page: Page,
	customElement: ElementHandle<SVGElement | HTMLElement> | null
) {
	return await page.evaluate((element) => {
		// Check if the custom element exists. If not, log an error message and return.
		const shadowRoot = element
			? (element as Element).shadowRoot
			: console.log("no element", element);
		if (!shadowRoot) {
			console.log("no shadowRoot", shadowRoot);
			return;
		}

		// Select all elements with the class "set-class.row._grid_columns" from the shadow root of the custom element.
		// Map over each selected element (row) and extract the symbol and name from the appropriate cells.
		// Return an array of objects, each containing the symbol and name from a row.
		return Array.from(
			shadowRoot.querySelectorAll("set-class.row._grid_columns")
		).map((row: any) => {
			const symbol_cell = row.querySelector("div._cell.symbol text-binding");
			const symbol_name_cell = row.querySelector(
				"div._cell.symbolName text-binding"
			);
			const symbol_text = symbol_cell
				? symbol_cell.shadowRoot.textContent
				: null;
			const name_text = symbol_name_cell
				? symbol_name_cell.shadowRoot.textContent
				: null;
			return { symbol: symbol_text, name: name_text };
		});
	}, customElement);
}

// This function is designed to filter an array of row objects into two categories based on the contents of the symbol and name properties.
function filterEquities(rows: any) {
	console.log("rows", rows);

	// Filter the rows into two categories: nonEquities and equities.
	// nonEquities contains all rows where the symbol includes "-PR", "-WT", or "-PF-", or the name includes "USD".
	// equities contains all rows that do not meet the criteria for nonEquities.

	// NoteToSelf: if performance is an issue, perhaps the following would be faster with
	// a Regex pattern.
	const nonEquities = rows.filter(
		(row: any) =>
			row.symbol.includes("-PR") ||
			row.symbol.includes("-WT") ||
			row.symbol.includes("-PF-") ||
			row.name.includes("USD")
	);
	const equities = rows.filter(
		(row: any) =>
			!row.symbol.includes("-PR") &&
			!row.symbol.includes("-WT") &&
			!row.symbol.includes("-PF-") &&
			!row.name.includes("USD")
	);
	// return { equities, nonEquities };
	return { equities };
}

// This function is designed to scroll through a webpage in a browser environment.
// It's necessary because barchart data tables do not fully load until they are in the viewport.
// By scrolling to the bottom of the page, we ensure that all elements have been loaded and displayed.
async function scrollPageToBottom(page: Page) {
	return await page.evaluate(() => {
		return new Promise((resolve) => {
			let totalHeight = 0; // This variable keeps track of the total height scrolled so far.
			const distance = 100; // This is the distance that will be scrolled in each interval.

			// We use setInterval to create a loop that scrolls the page and checks if we've reached the bottom.
			const timer = setInterval(() => {
				const scrollHeight = document.body.scrollHeight; // This is the total scrollable height of the body element.
				window.scrollBy(0, distance); // This line scrolls the window by the specified distance.
				totalHeight += distance; // We add the distance scrolled to the total height.

				// If the total height scrolled is greater than or equal to the scrollable height, we've reached the bottom.
				if (totalHeight >= scrollHeight) {
					clearInterval(timer); // We clear the interval to stop the loop.
					resolve(true); // We resolve the promise with 'true' to indicate that the function has completed successfully.
				}
			}, 100); // The loop runs every 100 milliseconds.
		});
	});
}

// This function is designed to automate the process of visiting a series of webpages,
// scrolling to the bottom of each page to ensure all content is loaded, and then
// extracting and processing data from each page.
export async function extractIndustryEquities(industries: any) {
	// Launch a new browser instance.
	const browser: ChromiumBrowser = await chromium.launch({ headless: false });
	// Create a new browser context. This represents a single session of browsing activity (including cookies, localStorage, etc).
	const context = await browser.newContext();

	let industryEquities: any = {};

	// Loop over each industry in the industries array.
	for (const industry of industries) {
		// Create a new page in the browser context.
		const page: Page = await context.newPage();
		// Construct the URL for the current industry.
		const url = `https://www.barchart.com/stocks/quotes/-${industry}/components?orderBy=weightedAlpha&orderDir=desc&page=all`;
		// Navigate to the constructed URL.
		await page.goto(url);
		// Scroll to the bottom of the page to ensure all content is loaded.
		await scrollPageToBottom(page);
		// Wait for the "bc-data-grid" element to be loaded in the DOM.
		await page.waitForSelector("bc-data-grid");

		// Select the "bc-data-grid" element from the page.
		const customElement: ElementHandle<SVGElement | HTMLElement> | null =
			await page.$("bc-data-grid");
		// Extract the rows of data from the selected element.
		const rows = await extractSymbolAndName(page, customElement);

		// If rows were successfully extracted...(incase of error)
		if (rows) {
			const { equities } = filterEquities(rows);
			industryEquities[industry] = { equities };
		}

		upload_to_db_barchart(industryEquities).catch((error) => {
			console.log(error);
		});

		// Close the current page.
		await page.close();
	}

	// Close the browser once all industries have been processed.
	await browser.close();

	return industryEquities;
}

File: src/data/barchart.py
from typing import Optional
from playwright.async_api import Page, ElementHandle
from playwright.async_api import async_playwright


async def extract_symbol_and_name(page: Page, custom_element: Optional[ElementHandle]) -> list:
    return await page.evaluate("""(element) => {
        const shadowRoot = element ? element.shadowRoot : console.log("no element", element);
        if (!shadowRoot) {
            console.log("no shadowRoot", shadowRoot);
            return;
        }
        return Array.from(shadowRoot.querySelectorAll("set-class.row._grid_columns")).map((row) => {
            const symbol_cell = row.querySelector("div._cell.symbol text-binding");
            const symbol_name_cell = row.querySelector("div._cell.symbolName text-binding");
            const symbol_text = symbol_cell ? symbol_cell.shadowRoot.textContent : null;
            const company_name_text = symbol_name_cell ? symbol_name_cell.shadowRoot.textContent : null;
            return { symbol: symbol_text, symbolName: company_name_text };
        });
    }""", custom_element)

def filter_equities(rows: list) -> dict:
    # This line creates a list of invalid symbols. It checks each row in the input list 'rows' and includes it in 'invalid_symbols' if the symbol contains "-PR", "-WT", "-PF-", or if the symbol name contains "USD".
    invalid_symbols = [row for row in rows if "-PR" in row['symbol'] or "-WT" in row['symbol'] or "-PF-" in row['symbol'] or "USD" in row['symbolName']]
    
    # This line creates a list of valid symbols. It checks each row in the input list 'rows' and includes it in 'valid_symbols' if the symbol does not contain "-PR", "-WT", "-PF-", "-B", "USD", or "Cl B".
    # It also modifies the symbol in each row based on certain conditions:
    # - If the symbol ends with '.VN', it replaces '.VN' with '.V'.
    # - If the symbol starts with '$', it removes the '$'.
    # - If the symbol starts with '-', it removes the '-'.
    valid_symbols = [dict(row, symbol=row['symbol'].replace('.VN', '.V')) if row['symbol'].endswith('.VN') else dict(row, symbol=row['symbol'].replace('$', '')) if row['symbol'].startswith('$') else dict(row, symbol=row['symbol'].lstrip('-')) if row['symbol'].startswith('-') else row for row in rows if not any(substring in row['symbol'] for substring in ["-PR", "-WT", "-PF-", "-B"]) and "USD" not in row['symbolName'] and "Cl B" not in row['symbolName']]
    
    # The function returns the list of valid symbols.
    return valid_symbols

async def scroll_page_to_bottom(page: Page):
    return await page.evaluate("""() => {
        return new Promise((resolve) => {
            let totalHeight = 0;
            const distance = 100;
            const timer = setInterval(() => {
                const scrollHeight = document.body.scrollHeight;
                window.scrollBy(0, distance);
                totalHeight += distance;
                if (totalHeight >= scrollHeight) {
                    clearInterval(timer);
                    resolve(true);
                }
            }, 100);
        });
    }""")
async def scroll_page_to_top(page: Page):
    return await page.evaluate("""() => {
        return new Promise((resolve) => {
            const distance = 100;
            const timer = setInterval(() => {
                if (window.scrollY > 0) {
                    window.scrollBy(0, -distance);
                } else {
                    clearInterval(timer);
                    resolve(true);
                }
            }, 100);
        });
    }""")

async def extract_data(url, context, page: Optional[Page] = None):
    if page is None:
        page = await context.new_page()
        await page.goto(url)
    await scroll_page_to_bottom(page)
    await page.wait_for_selector("bc-data-grid")

    custom_element = await page.query_selector("bc-data-grid")
    rows = await extract_symbol_and_name(page, custom_element)

    await page.close()

    return rows

async def extract_sectors(context):
    # object that gets returned
    sectors = {}

    url = "https://www.barchart.com/search?q=$TT&regions=ca&assets=indices"
    rows = await extract_data(url, context)

    if rows:
        filtered_rows = filter_equities(rows)
        for row in filtered_rows:
            symbol = row['symbol']
            symbol_name = row['symbolName']
            sectors[symbol] = symbol_name

    return sectors

async def extract_industries(context):
    # object that gets returned
    industries = {}

    url = "https://www.barchart.com/ca/stocks/sectors/industry-performance?orderBy=priceChange&orderDir=desc"
    page = await context.new_page()
    await page.goto(url)

    # Uncheck the checkbox
    await page.evaluate("""() => {
        let checkbox = document.querySelector('#show-industrial');
        if (checkbox.checked) {
            checkbox.click();
        }
    }""")

    await scroll_page_to_bottom(page)
    await page.click('.show-all.ng-scope')
    await scroll_page_to_top(page)

    rows = await extract_data(url, context, page)

    if rows:
        filtered_rows = filter_equities(rows)
        for row in filtered_rows:
            symbol = row['symbol']
            symbol_name = row['symbolName']
            industries[symbol] = symbol_name

    return industries

async def extract_industry_equities(industries, context):
    # object that gets returned
    industry_equities = {}

    # If industries is a string, make it a list
    if isinstance(industries, str):
        industries = [industries]

    for industry in industries.keys():
        url = f"https://www.barchart.com/stocks/quotes/-{industry}/components?orderBy=weightedAlpha&orderDir=desc&page=all"
        equities = await extract_data(url, context)

        if equities:
            equities = filter_equities(equities)
            industry_equities[industry] = equities

    return industry_equities

async def setup_context():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=False)
        context = await browser.new_context()
        return context, browser

